internal/gitstream/assign.go

```diff
@@ -78,49 +80,44 @@ func (a *Assign) assignIssues(ctx context.Context) error {
 			return fmt.Errorf("error while looking for SHAs in %q: %v", *issue.Body, err)
 		}
 
+		commitAuthors := make([]*github.User, 0, len(shas))
 		for _, s := range shas {
-
-			if a.DryRun {
-				logger.Info("Dry run: skipping issue update")
-				return nil
-			}
-
-			logger.Info("Assigning issue")
-
-			var (
-				assignee string
-				intErr   error
-			)
+			logger = logger.WithValues("commit", s.String())
 			if user, err := a.UserHelper.GetCommitAuthor(ctx, s.String()); err != nil {
 				if !errors.Is(err, gh.ErrUnexpectedReply) {
-					logger.Info("WARNING: failed to get a response from Github, skipping commit",
-						"issue", *issue.Number, "error", err.Error())
-					continue
-				}
-				logger.Info("WARNING: commit author for downstream issue not found on github, picking a random assignee",
-					"issue", *issue.Number, "error", err.Error())
-				assignee, intErr = a.OwnersHelper.GetRandomApprover(owners)
-				if intErr != nil {
-					return fmt.Errorf("could not get a random owner: %v", err)
+					logger.Info("WARNING: failed to get commit author from Github", "error", err.Error())
+				} else {
+					multiErr = multierror.Append(multiErr, fmt.Errorf("failed to get commit author from Github: %v", err))
 				}
 			} else {
-				if a.OwnersHelper.IsApprover(owners, *user.Login) {
-					assignee = *user.Login
+				commitAuthors = append(commitAuthors, user)
+			}
+		}
+
+		var assigned bool
+		var errFailedToAssignIssue = errors.New("could not assign issue")
+		for _, ca := range commitAuthors {
+			if authorLogin := *ca.Login; a.OwnersHelper.IsApprover(owners, authorLogin) {
+				logger.Info("Assigning issue", "author", authorLogin)
+				if err := a.IssueHelper.Assign(ctx, issue, authorLogin); err != nil {
+					multiErr = multierror.Append(multiErr, errFailedToAssignIssue)
 				} else {
-					logger.Info("commit author for downstream issue is not an owner, picking a random assignee",
-						"issue", *issue.Number)
-					assignee, intErr = a.OwnersHelper.GetRandomApprover(owners)
-					if intErr != nil {
-						return fmt.Errorf("could not get a random owner: %v", err)
-					}
+					assigned = true
 				}
 			}
+		}
 
-			if err := a.IssueHelper.Assign(ctx, issue, assignee); err != nil {
+		// We should not pick a random approver if we failed to assign each one of the commit authors
+		if !assigned && !errors.Is(multiErr, errFailedToAssignIssue) {
```

Comment:
We never reset `multiErr` in this function, so if a previous issue / PR added `errFailedToAssignIssue` to it, then that part of the `if` will always be `true`.

Reply:
If we find a way to create one `multiErr` per issue / PR, then we might get rid of the `assigned` boolean.

Reply:
The issue with resetting it means that the function will return an error only if the last issue had any error, while we are trying to capture every errors in each one of the issues.

Reply:
This is exactly what I did. Check the new code.